# Objetivos 
El coranavirus condiciono a todos a aislarse de los demas lo que conllevo a dejar de hacer actividades al aire libre, esto conllevo a centrarse en actividades que se pudieran realizar sin la necesidad de depender de alguien que este fisicamente presente, gracias a esto las startups vieron una opurtunidad con los amantes de los libros, conbinando con esto con enfoque tecnologico se apresuron a realizar apliciones para este nicho.

Nos han dado una base de datos que contempla los datos de unos de estas aplicaciones de la cual se no pidio analizarla para crear una propuesta de valor diferencial

In [1]:

import pandas as pd
from sqlalchemy import create_engine
from datetime import date

## Conexion a BD

In [2]:
db_config = { 'user': 'practicum_student', # username
         'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # password
         'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
         'port': 5432, # connection port
         'db': 'data-analyst-final-project-db' # the name of the database
     }
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],db_config['pwd'],db_config['host'],db_config['port'],db_config['db'])
    
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
def query_result(query):

    db_config = { 'user': 'practicum_student', # username
         'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # password
         'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
         'port': 5432, # connection port
         'db': 'data-analyst-final-project-db' # the name of the database
     }
    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],db_config['pwd'],db_config['host'],db_config['port'],
                                                         db_config['db'])
    
    engine = create_engine(connection_string, connect_args={'sslmode':'require'})
    return pd.io.sql.read_sql(query, con = engine)

## Estudio de tablas

### Books

In [4]:
query_result('SELECT * FROM books LIMIT 5')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query_result('SELECT COUNT(*) FROM books WHERE book_id IS NULL')

count
0      0

In [6]:
query_result('SELECT COUNT(*) FROM books')

count
0   1000

In [7]:
#verificar si existen libros duplicados buscando por su id y titulo
query_result('SELECT book_id,title FROM books GROUP BY book_id,title  HAVING COUNT(1) > 1 ')

Empty DataFrame
Columns: [book_id, title]
Index: []

-La tabla books cuenta con registros de las caracteristicas de los libros su nombre, autor, fecha de publicacion etc

-La tabla no tiene valores nulos y cuenta con 1000 registros

-tampoco se encontraron registros duplicados

### Authors

In [8]:
query_result('SELECT * FROM authors LIMIT 5')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
query_result('SELECT * FROM authors WHERE author_id IS NULL')

Empty DataFrame
Columns: [author_id, author]
Index: []

In [10]:
query_result('SELECT COUNT(*) FROM authors ')

count
0    636

In [11]:
query_result('SELECT author_id,author FROM authors GROUP BY author_id,author HAVING COUNT(1)>1')

Empty DataFrame
Columns: [author_id, author]
Index: []

-La tabla autores solo cuenta con el nombre del autor y su identidicador unico

-La tabla no tiene registros nulos y cuenta con 636 registros

-Tampoco se cuenta con registros duplicados

### Ratings

In [12]:
query_result('SELECT * FROM ratings LIMIT 5')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
query_result('SELECT * FROM ratings WHERE rating_id IS NULL')

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

In [14]:
query_result('SELECT COUNT(*) FROM ratings')

count
0   6456

In [15]:
query_result('SELECT rating_id,book_id FROM ratings GROUP BY rating_id,book_id HAVING COUNT(1) > 1')

Empty DataFrame
Columns: [rating_id, book_id]
Index: []

-La tabla ratings cuenta con el nombre de usuario de la persona dio el rating, la informacion unica del libro y su respectivo rating

-La tabla no tiene registros nulos y cuenta con 6456 registros

-No posee registros duplicados

### Reviews

In [16]:
query_result('SELECT * FROM reviews LIMIT 5')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [17]:
query_result('SELECT * FROM reviews WHERE review_id IS NULL')

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

In [18]:
query_result('SELECT COUNT(*) FROM reviews ')

count
0   2793

In [19]:
query_result('SELECT review_id, book_id FROM reviews GROUP BY review_id, book_id HAVING COUNT(1) > 1')

Empty DataFrame
Columns: [review_id, book_id]
Index: []

-La tabla reviews cuenta con la reseña textual, el usuario quien la redacto y sus respectivos identifacadores unicos de libro y la reviews misma

-La tabla no cuenta con resgristros nulos y tiene 2793 registros

-No posee registros duplicados

### Publishers

In [20]:
query_result('SELECT * FROM publishers LIMIT 5')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [21]:
query_result('SELECT * FROM publishers WHERE publisher_id IS NULL')

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

In [22]:
query_result('SELECT COUNT(*) FROM publishers')

count
0    340

In [23]:
query_result('SELECT publisher_id,publisher FROM publishers GROUP BY publisher_id,publisher HAVING COUNT(1)>1')

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

-La tabla publishers cuenta con el nombre de la editorial y su identificador unico

-La tabla no posee registros nulos y cuenta con 340 registros

-No posee registros duiplicados

## Tareas

### Encuentra el número de libros publicados después del 1 de enero de 2000.

In [24]:
query_result(f"""SELECT COUNT(*) AS cntl FROM books WHERE publication_date > '2000-01-01'""")

cntl
0   819

-Existe 819 libros que se publicaron despues del 1 de Enero del 2000 esto es un 81.9% de los registros por lo que la gran mayoria son libros de hace aproximadamente 24 años

### Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [25]:
query_result(f"""SELECT books.book_id ,
                                books.title as title,
                                COUNT (DISTINCT reviews.review_id) AS cnt_re,
                                ROUND(AVG(ratings.rating),2) AS avg_ra
                            FROM books 
                            LEFT JOIN reviews ON books.book_id  = reviews.book_id
                            LEFT JOIN ratings ON books.book_id  = ratings.book_id 
                            GROUP BY books.book_id,books.title
                            ORDER BY avg_ra DESC
                            LIMIT 10""")

book_id                                              title  cnt_re  avg_ra
0       86      Arrows of the Queen (Heralds of Valdemar  #1)       2     5.0
1      169  Crucial Conversations: Tools for Talking When ...       2     5.0
2       62                                      Alas  Babylon       2     5.0
3       57                    Act of Treason (Mitch Rapp  #9)       2     5.0
4       76                                        Angels Fall       2     5.0
5      136  Captivating: Unveiling the Mystery of a Woman'...       2     5.0
6       55          A Woman of Substance (Emma Harte Saga #1)       2     5.0
7       20              A Fistful of Charms (The Hollows  #4)       2     5.0
8       17                      A Dirty Job (Grim Reaper  #1)       4     5.0
9      182                                         Dead Souls       2     5.0

In [26]:
query_result(f"""SELECT books.book_id ,
                                books.title as title,
                                COUNT (DISTINCT reviews.review_id) AS cnt_re,
                                ROUND(AVG(ratings.rating),2) AS avg_ra
                            FROM books 
                            LEFT JOIN reviews ON books.book_id  = reviews.book_id
                            LEFT JOIN ratings ON books.book_id  = ratings.book_id 
                            GROUP BY books.book_id,books.title
                            ORDER BY avg_ra ASC
                            LIMIT 10""")

book_id                                              title  cnt_re  avg_ra
0      303                               Harvesting the Heart       2    1.50
1      371                                              Junky       2    2.00
2      202                                      Drowning Ruth       3    2.00
3      316                  His Excellency: George Washington       2    2.00
4      915  The World Is Flat: A Brief History of the Twen...       3    2.25
5      794                                  The Mermaid Chair       3    2.33
6      772                             The Kitchen God's Wife       3    2.33
7      313              Heretics of Dune (Dune Chronicles #5)       2    2.50
8       72                                          Amsterdam       2    2.50
9      193                                        Dr. Faustus       2    2.50

In [27]:
q=f"""SELECT books.book_id ,
                                books.title as title,
                                COUNT (DISTINCT reviews.review_id) AS cnt_re,
                                ROUND(AVG(ratings.rating),2) AS avg_ra
                            FROM books 
                            LEFT JOIN reviews ON books.book_id  = reviews.book_id
                            LEFT JOIN ratings ON books.book_id  = ratings.book_id 
                            GROUP BY books.book_id,books.title
                            ORDER BY avg_ra 
                            """

df_re_ra = pd.read_sql_query(q, con = engine)

In [28]:
print(df_re_ra)

     book_id                                              title  cnt_re  \
0        303                               Harvesting the Heart       2   
1        371                                              Junky       2   
2        316                  His Excellency: George Washington       2   
3        202                                      Drowning Ruth       3   
4        915  The World Is Flat: A Brief History of the Twen...       3   
..       ...                                                ...     ...   
995      421                                        Marvel 1602       2   
996      418                                              March       2   
997      169  Crucial Conversations: Tools for Talking When ...       2   
998      901  The Walking Dead  Book One (The Walking Dead #...       2   
999      902  The War of Art: Break Through the Blocks & Win...       2   

     avg_ra  
0      1.50  
1      2.00  
2      2.00  
3      2.00  
4      2.25  
..      ...  
9

In [29]:
print(df_re_ra.groupby('cnt_re').agg({'cnt_re':'count','avg_ra':'mean'}))

        cnt_re    avg_ra
cnt_re                  
0            6  3.945000
1           37  3.815135
2          444  3.914369
3          277  3.862238
4          160  3.908375
5           60  3.957500
6           15  3.996000
7            1  3.660000


In [30]:
print(df_re_ra[df_re_ra['avg_ra']>=3])

     book_id                                              title  cnt_re  \
35       876  The Stranger Beside Me: Ted Bundy: The Shockin...       2   
36       877                                  The Street Lawyer       3   
37       585              Son of a Witch (The Wicked Years  #2)       2   
38       178                             Dark Prince (Dark  #1)       2   
39       694                               The Crying of Lot 49       2   
..       ...                                                ...     ...   
995      421                                        Marvel 1602       2   
996      418                                              March       2   
997      169  Crucial Conversations: Tools for Talking When ...       2   
998      901  The Walking Dead  Book One (The Walking Dead #...       2   
999      902  The War of Art: Break Through the Blocks & Win...       2   

     avg_ra  
35      3.0  
36      3.0  
37      3.0  
38      3.0  
39      3.0  
..      ...  
9

In [31]:
print(df_re_ra[df_re_ra['avg_ra']<3])

    book_id                                              title  cnt_re  avg_ra
0       303                               Harvesting the Heart       2    1.50
1       371                                              Junky       2    2.00
2       316                  His Excellency: George Washington       2    2.00
3       202                                      Drowning Ruth       3    2.00
4       915  The World Is Flat: A Brief History of the Twen...       3    2.25
5       794                                  The Mermaid Chair       3    2.33
6       772                             The Kitchen God's Wife       3    2.33
7       822                                     The Penelopiad       2    2.50
8       412                   Lucy Sullivan Is Getting Married       2    2.50
9       747         The High Lord (Black Magician Trilogy  #3)       2    2.50
10      929                                      Thunderstruck       2    2.50
11      313              Heretics of Dune (Dune Chro

-La mayor cantidad de reseñas por libros es 2 y la mayor es de 7, se puede ver que no se han hecho reseñas de todos los libros.

-El rating de un libro que se puede decir que es bueno se dejara en este caso de 3 hacia arriba los cuales son la gran mayoria.

-Con esto podemos concluir que los libros no estan teniendo muchas reseñas y que por lo general se califican como bastante buenas.

### Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas.

In [32]:
query_result(f"""SELECT publishers.publisher_id,
                                            publishers.publisher,
                                            COUNT(books.publisher_id) AS cnt_publ
                                    FROM
                                        books
                                        INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
                                    WHERE books.num_pages > 50
                                    GROUP BY publishers.publisher_id,publishers.publisher
                                    ORDER BY cnt_publ DESC
                                    LIMIT 10""")

publisher_id                 publisher  cnt_publ
0           212             Penguin Books        42
1           309                   Vintage        31
2           116  Grand Central Publishing        25
3           217          Penguin Classics        24
4            33          Ballantine Books        19
5            35                    Bantam        19
6            45                   Berkley        17
7           284        St. Martin's Press        14
8            46             Berkley Books        14
9            83                     Delta        13

-La editorial Penguin Books es la que mas libros ha publicado con 42, sabiendo esto se puede buscar hacer una asociacion con la editora para realizar publicaciones dentro de la aplicacion

### Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [33]:
query_result(f"""SELECT authors.author_id,
                        authors.author as name_author,
                       
                        AVG(book_avg_ratings.avg_rating) as a_avg_rating
                 FROM authors 
                 JOIN books ON authors.author_id = books.book_id
                 JOIN(
                     SELECT book_id,
                             AVG(rating) AS avg_rating
                     FROM ratings
                     GROUP BY book_id
                     HAVING COUNT(rating)>=50) AS book_avg_ratings ON books.book_id = book_avg_ratings.book_id
                 GROUP BY authors.author_id,authors.author
                 ORDER BY a_avg_rating DESC""")


author_id                                        name_author  a_avg_rating
0         302                                       John Grisham      4.414634
1         299                       John Eldredge/Stasi Eldredge      4.287500
2         300                                        John Fowles      4.246575
3         399                   Mark Twain/Shelly Fisher Fishkin      4.192308
4         301                                          John Gray      4.186667
5         405                                          Mary Karr      3.901408
6         627  William Shakespeare/Barbara A. Mowat/Paul Wers...      3.789474
7         545               Stephen King/Bettina Blanch Tyroller      3.787879
8          79                                        Chaim Potok      3.729730
9          75                                 Caroline B. Cooney      3.678571
10        488                                   Raymond E. Feist      3.622951

In [48]:
query_result(f"""SELECT authors.author,authors.author_id,
                        book_ra.avg_ra
                        FROM authors
                        JOIN (
                                SELECT books.author_id AS author_id,
                                       ROUND(AVG(ratings.rating),2 )AS avg_ra
                                FROM books
                                JOIN ratings ON books.book_id = ratings.book_id
                                GROUP BY books.author_id
                                HAVING COUNT(ratings.book_id) > 50  
                             ) AS book_ra ON authors.author_id = book_ra.author_id
                        ORDER BY book_ra.avg_ra DESC
                             
                             
                             """)


author  author_id  avg_ra
0                          J.K. Rowling/Mary GrandPré        236    4.29
1                                     Agatha Christie          3    4.28
2                   Markus Zusak/Cao Xuân Việt Khương        402    4.26
3                                      J.R.R. Tolkien        240    4.24
4                            Roald Dahl/Quentin Blake        499    4.21
5                                   Louisa May Alcott        376    4.20
6                                        Rick Riordan        498    4.13
7                                       Arthur Golden         39    4.11
8                                        Stephen King        542    4.01
9                                        John Grisham        302    3.97
10                                    William Golding        621    3.90
11                                       Jodi Picoult        291    3.88
12                                    Sophie Kinsella        536    3.88
13                                    Nicholas Sparks        445    3.88
14                                    James Patterson        257    3.86
15                                      J.D. Salinger        235    3.85
16  William Shakespeare/Paul Werstine/Barbara A. M...        630    3.79
17           Paulo Coelho/Alan R. Clarke/Özdemir İnce        469    3.79
18                                         Lois Lowry        372    3.74
19                                          Dan Brown        106    3.74
20            George Orwell/Boris Grabnar/Peter Škerl        195    3.73
21                                    Stephenie Meyer        554    3.66
22                                     John Steinbeck        311    3.64

In [67]:
query_result(f"""SELECT  books.author_id AS author_id,
                                       AVG(ratings.rating) AS avg_ra,
                                       COUNT(ratings.book_id) 
                                FROM books
                                JOIN ratings ON books.book_id = ratings.book_id
                                GROUP BY books.author_id
                                HAVING COUNT(ratings.book_id) > 50
                                
                                ORDER BY COUNT(ratings.book_id) DESC
                             
                             """)


author_id    avg_ra  count
0         236  4.288462    312
1         240  4.240964    166
2         554  3.662500    160
3         106  3.741259    143
4         445  3.882883    111
5         542  4.009434    106
6         235  3.846939     98
7         498  4.130952     84
8         291  3.881579     76
9         195  3.729730     74
10        311  3.643836     73
11        621  3.901408     71
12        302  3.971429     70
13        630  3.787879     66
14        372  3.738462     65
15        257  3.859375     64
16        499  4.209677     62
17        536  3.877193     57
18        469  3.789474     57
19         39  4.107143     56
20        376  4.203704     54
21          3  4.283019     53
22        402  4.264151     53

In [62]:
query_result(f"""SELECT books.book_id, books.author_id AS author_id,
                                       AVG(ratings.rating) AS avg_ra,
                                       COUNT(ratings.book_id)
                                FROM books
                                JOIN ratings ON books.book_id = ratings.book_id
                                WHERE books.author_id =236
                                
                                GROUP BY books.book_id,books.author_id
                               HAVING COUNT(ratings.book_id)>50
                             
                             """)


book_id  author_id    avg_ra  count
0      299        236  4.287500     80
1      300        236  4.246575     73
2      301        236  4.186667     75
3      302        236  4.414634     82

-Las autoras J.K. Rowling/Mary GrandPré	son las que poseen el promedio mas alto de calificaciones, como se puede apreciar esto es debido a que poseen varios libros, en este caso una saga Harry Potter la cual tiene una gran cantidad de calificaciones sus diferentes volumenes por lo que enfocarse en libros que posean varios volumenes es una buena idea.

### Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [110]:
query_result(f""" WITH user_review_counts AS (
    SELECT 
        r.username AS user,
        COUNT(r.review_id) AS review_count
    FROM 
        reviews r
    WHERE 
        r.username IN (
                        SELECT 
                            rt.username
                        FROM 
                            ratings rt
                        GROUP BY 
                            rt.username
                        HAVING 
                            COUNT(rt.book_id) > 50
                    )
    GROUP BY 
        r.username
)

SELECT 
    user_review_counts.user,
    user_review_counts.review_count,
    (SELECT AVG(review_count) FROM user_review_counts) AS avg_text_reviews
FROM 
    user_review_counts
ORDER BY 
    avg_text_reviews DESC;






""")

user  review_count  avg_text_reviews
0     sfitzgerald            28         24.333333
1  jennifermiller            25         24.333333
2          xdavis            18         24.333333
3          paul88            22         24.333333
4      martinadam            27         24.333333
5       richard89            26         24.333333

-El promedio de reseñas de texto que tienen los usuarios que hicieron mas de 50 calificaciones es de 24 por lo que se podria
toma este numero como referencia base para ver si un libro es exitoso ya que llamo la suficiente atención para que la personas
lo reseñen